content/blog/2019-12-18-dfg-cover/index.md

```diff
@@ -0,0 +1,383 @@
+
+[embench]: https://embench.org
+[`networkx.isomorphism`]: https://networkx.github.io/documentation/stable/reference/algorithms/isomorphism.html
+
+## Identifying common DFG stencils
+
+Of course, finding the common subgraphs by hand is pretty antithetical to a
+reasonable approach at compiling code. Our real goal is to automate the process
+of finding the common DFG stencils to accelerate.
+
+### Formal description of the task
+
+In this context of ignoring control flow and considering data flows within basic
+block, we can look at the problem purely graph-theoretically. For a single trace
+through the program, the data flow graph $G$ is acyclic, and we would like to
+cover as much of it as possible with subgraphs corresponding to the stencils
+that we accelerate. Statically, we do not know what the final data flow graph
+is, but we do know that we will be able to assemble one by connecting dangling
+edges from control-flow-free components: basic blocks.
+
+We would like to find a small collection of graph components $\mathcal H = \{H_i, \ldots, H_k\}$, which we can use to replace parts of and accelerate programs having basic blocks $\mathcal G = \{G_1,\ldots,G_n\}$, that maximizes the total saved time:
+
+$$\mathcal S_{\mathcal H}(\mathcal G) := \max_{\mathcal C \in \text{Cov}(\mathcal G, \mathcal H)}~ \sum_{G \in \mathcal G} w_G \cdot  \sum_{H \in \mathcal C_G} f_H \cdot |H|$$
+
+where:
+
+- $\text{Cov}(\mathcal G, \mathcal H)$ is the set of all valid (partial)
+    coverings of basic blocks with at most one stencil, that is, injective graph
+     morphisms $\varphi: (\cup \mathcal G) \to \cup \mathcal H$.
+- $\mathcal C_G$ is the component of the covering $\mathcal C$ of the
+    total covering on the particular basic block graph $G$.
+- $w_G$ is independent of $\mathcal H$ and proportional to the expected
+    number of times $G$ is executed.
+- $f_H$ is the expected speedup factor from accelerating the component $H$.
+
+To consider this as a general learning problem, imagine that there's some
+underlying distribution $\mathtt{Programs}$ of programs that people write; we
+can now cast our work as a solution to the optimization problem of finding:
+
+$$ \arg\max_{\mathcal H}\left( \mathop{\mathbb E}\limits_{\mathcal G\sim \texttt{Programs}}~ \mathcal S_{\mathcal H}(\mathcal G) - \text{Cost}(\mathcal H) \right)$$
+
+where $\text{Cost}(\mathcal H)$ is the additional cost incurred by choosing to
+accelerate the subgraph stencil $\mathcal H$, which is higher for larger
+subgraphs. In this learning analogy, finding common DFGs for a particular
+collection of programs is training data.
+
+Rather than solve this optimization problem in closed form, we optimize for
+the heuristics of finding candidate subgraphs based on specifying (1) the size
+of each subgraph, and (2) the number of subgraphs we can choose. These two
+heuristics are effectively regularization knobs that could be used in future
+work to automate the entire optimization.
```

Comment:
Huh, I also don't get the connection between these parameters and regularization. Is it really regularization if you're just simplifying the problem (i.e., making it tractable) by making some assumptions? Is there some danger of overfitting that I'm missing here?

Reply:
We're regularizing the search space by artificially imposing a cost for larger or more sub-graphs. Just like $\ell^1$ regularization, we are explicitly adding a preference for short descriptions, to avoid overfitting to a single program. The example of overfitting is the one given above: if we just set $\mathcal H := \mathcal G$, then we've overfit to the particular program we were using to try to infer general stencils.

Reply:
Excellent; this also helps me clarify. An overfit solution would just encode exactly what the program wants to do as one giant instruction. Makes sense!